### PR TITLE
fix: started on Add duration option #4

### LINKED ISSUE #4: Add duration option
Make it possible to tweak the time before the snow ends. It is now hardcoded to 15s.

### PR DESCRIPTION
**Pull request by olavea**
It's not working right, yet. I'll look more at it tomorrow

### CODE REVIEW

**Comment by olavea** [comment]
I am looking at this Gatsby doc here to see how 

`exports.pluginOptionsSchema = ({ Joi }) => {`
 
is different from 
`exports.createPages = async ({ graphql }) => {`

[gatsby-node/#async-vs-sync-work](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#async-vs-sync-work)

Question 1.

Our Let It Snow ❄️  plugin does not do async work, so we can return directly? Is that correct?

**Comment by raae** [comment]
Documentation on how to require a number with Joi: https://joi.dev/api/?v=17.4.2#number
It should also probably be required to be larger than 5s: https://joi.dev/api/?v=17.4.2
It should probably also be required to be an integer: https://joi.dev/api/?v=17.4.2#numberinteger

**Comment by raae** [comment]
> I am looking at this Gatsby doc here to see how
> 
> `exports.pluginOptionsSchema = ({ Joi }) => {`
> 
> is different from `exports.createPages = async ({ graphql }) => {`
> 
> [gatsby-node/#async-vs-sync-work](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#async-vs-sync-work)
> 
> Question 1.
> 
> Our Let It Snow ❄️ plugin does not do async work, so we can return directly? Is that correct?

`exports.pluginOptionsSchema` does not do any async work, and therefore does not need to be async. 

The actual work of the plugin is done in `gatsby-browser.js` inside `export const onInitialClientRender` also not async work.

**Comment by olavea** [comment]
@raae said look at @inezabonte 's code and I think you will find the answer. Only difference is you need numbers, while Ineza  


Question 2.

**Comment by olavea** [comment]
I did it in
```
// gatsby-node.js
    duration: Joi.array()
      .items(Joi.number().min(3).max(30).integer().required().

```
Thes stuff you asked for here:
Documentation on how to require a number with Joi: https://joi.dev/api/?v=17.4.2#number
It should also probably be required to be larger than 5s: https://joi.dev/api/?v=17.4.2
It should probably also be required to be an integer: https://joi.dev/api/?v=17.4.2#numberinteger

**Comment by olavea** [comment]
Where is the Pirate Princess supposed to choose when the snow should stop in the ui?

**Comment by raae** [comment]
> Where is the Pirate Princess supposed to choose when the snow should stop in the ui?

There is no UI for it, duration is a plugin option and is configured in the gatsby-config.js of the project using the plugin. You can test it by running the demo project included.

File: plugin/gatsby-node.js [hunk 1]
```diff
@@ -4,5 +4,13 @@ exports.pluginOptionsSchema = ({ Joi }) => {
     colors: Joi.array()
       .items(Joi.string().default("#ffffff"))
       .description("Array of hex color values"),
+    duration: Joi.array()
```
Comment (raae) [inline plugin/gatsby-node.js]: **issue:** Duration is not an array of numbers, it's a number.

Comment (olavea) [reply plugin/gatsby-node.js]:
```js
/gatsby-node/
exports.pluginOptionsSchema = ({ Joi }) => {
  return Joi.object({
    ....
    duration: Joi.number()
  });
};
```

**Comment by olavea** [comment]
I'll do something more tomorrow 😸 👍

File: plugin/gatsby-node.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ exports.pluginOptionsSchema = ({ Joi }) => {
     intensity: Joi.string()
       .valid("regular", "light", "blizzard")
       .default("regular"),
```
Comment (olavea) [inline plugin/gatsby-node.js:11]: This works, but am I done?
What more do I need to do?

File: plugin/gatsby-browser.js [hunk 3]
```diff
@@ -3,9 +3,9 @@
 import confetti from "canvas-confetti";
 
 export const onInitialClientRender = (_, options) => {
-  const { colors = ["#ffffff"] } = options;
+  const { colors = ["#ffffff"], intensity, duration } = options;
```
Comment (raae) [inline plugin/gatsby-browser.js]: You also need to give duration a default of 15 seconds here.

Why, as is should come through the plugin schema set up? The schema code in gatsby-node is skipped in Gatsby v2 and we are supporting v2.

**Review by raae — CHANGES_REQUESTED** [review]

File: plugin/gatsby-browser.js (hunk 3 above)
Comment (raae) [reply plugin/gatsby-browser.js]: You so so by adding "= 15 * 1000" after duration.

Comment (raae) [reply plugin/gatsby-browser.js]: Never mind, we'll make Gatsby v2.4.0 required.

**Comment by raae** [comment]
Merge fix

File: plugin/gatsby-browser.js (hunk 3 above)
Comment (olavea) [reply plugin/gatsby-browser.js]: Waaait? I a NOT doing `= 15 * 1000`
😸

Comment (raae) [reply plugin/gatsby-browser.js]: No need, we are requiring version 2.4 that has plugin schema, so we are sure to get a default value 🎉

**Pull request merged**